### PR TITLE
Feature/hdf5 dump

### PR DESCRIPTION
**Pull request by RChrHill**
Initial implementation of a feature to combine many contractions into a single HDF5 file for output. This initial implementation is a proof-of-concept that works only with the Meson contractor, with minimal other changes. The final feature should be applicable for all contraction types - although the final output structure in the HDF5 file needs consideration.

### CODE REVIEW

File: Hadrons/Modules/MContraction/Meson.hpp [hunk 1]
```diff
@@ -85,6 +85,8 @@ class TMeson: public Module<MesonPar>
                                         Gamma::Algebra, gamma_snk,
                                         Gamma::Algebra, gamma_src,
                                         std::vector<Complex>, corr);
+        Result() {};
+        Result(uint16_t dummy) {};
```
Comment (RChrHill) [inline Hadrons/Modules/MContraction/Meson.hpp:89]: Dummy single-arg constructor required here due to the use of `envCreate` in `setup`

Comment (aportelli) [reply Hadrons/Modules/MContraction/Meson.hpp:89]: `enCreate` should not for you to do that, the variadic macro argument can be empty e.g. `MACRO(a,b,)`

File: Hadrons/Modules/MContraction/Meson.hpp [hunk 2]
```diff
@@ -153,14 +155,20 @@ template <typename FImpl1, typename FImpl2>
 std::vector<std::string> TMeson<FImpl1, FImpl2>::getOutput(void)
 {
     std::vector<std::string> output = {};
+    if (!par().output.length())
```
Comment (aportelli) [inline Hadrons/Modules/MContraction/Meson.hpp:158]: The logic is ok but somehow different compared to the rest of Hadrons. I would prefer the output to be saved in the environement in any case (it is very small), and that it is saved on disk if `par().output` is not empty.

File: Hadrons/Modules/MContraction/Meson.hpp [hunk 3]
```diff
@@ -249,7 +262,17 @@ void TMeson<FImpl1, FImpl2>::execute(void)
             }
         }
     }
-    saveResult(par().output, "meson", result);
+
+    
+    if (par().output.length())
```
Comment (aportelli) [inline Hadrons/Modules/MContraction/Meson.hpp:267]: I would prefer the boolean `empty()` to be used

**Review by aportelli — CHANGES_REQUESTED** [review]
Hi Ryan,

Thanks I made minor comment you can start implementing.

The overall strategy does work fine but would need to be statically instantiated for every possible module output. This will increase manual code maintenance and build times. It might be that we do not have a choice, however I am investigating a possible way to have a serialisable type able to write its HDF5 content regardless of the type itself. That would allow us to have a generic way of aggregating results. I'll let you know more about that later this week.

**Review by aportelli — APPROVED** [review]
Hi Felix, not clear if still critical but it can be useful and I am happy with the code so I'll just merge it for now.

File: Hadrons/Modules/MContraction/Meson.hpp (hunk 3 above)
Comment (RChrHill) [reply Hadrons/Modules/MContraction/Meson.hpp]: Now updated to use `empty()` in the two places where `length()` was previously used

File: Hadrons/Modules/MContraction/Meson.hpp (hunk 1 above)
Comment (RChrHill) [reply Hadrons/Modules/MContraction/Meson.hpp]: Constructors removed and macro called fixed in 6a34ab619a5d7cca1bb6abcaf3aac014e8fba517

**Comment by aportelli** [comment]
to be resubmitted

**Pull request closed without merging**